pkg/pillar/hardware/smart.go

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2022. Zededa, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package hardware
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"github.com/lf-edge/eve/api/go/info"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+var (
+	maxSmartCtlSize        = 65536
+	idCurrentPendingSector = 197
+	idPowerCycleCount      = 12
+	idPowerOnHours         = 9
+	idReallocatedSectorCt  = 5
+	diskInfoDir            = "/persist/disks_info"
+	pathDisksList          = "/persist/disks_info/SMART_device.json"
+	cmd                    = "smartctl"
+)
+
+// getPathForFileDisk return path for JSON file
+// with S.M.A.R.T info for diskName
+//
+// "/dev/sda" -> "/persist/disks_info/-dev-sda.json"
+func getPathForFileDisk(diskName string) string {
+	return filepath.Join(diskInfoDir,
+		strings.Replace(fmt.Sprintf("%s.json", diskName), "/", "-", -1))
+}
+
+func smartctlExec(filePath string, args ...string) error {
+	jsonData, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("run command %s %v failed %v", cmd, args, err)
+	}
+
+	file, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("unable to create file: %v, err: %v", filePath, err)
+	}
+	defer file.Close()
+
+	_, err = file.WriteString(string(jsonData))
+	if err != nil {
+		return fmt.Errorf("unable to write data in file: %v, err: %v", filePath, err)
+	}
+
+	return nil
+}
+
+// readFWithMaxSize returns the content but limits the size to maxReadSize
+func readFWithMaxSize(filename string, maxReadSize int) ([]byte, error) {
```

Comment:
We already have such a function in utils package called ReadWithMaxSize(). You should be able to use it and not define a new function.